### PR TITLE
Add two phase retrieval merge function [DPP-718]

### PR DESCRIPTION
**Pull request by nmarton-da**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by rautenrieth-da — APPROVED** [review]
The logic is quite involved, if this is critical code then I think it needs more tests for edge cases and more precise comments. Otherwise it looks good, I have left a few cosmetic comments.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala [hunk 1]
```diff
@@ -135,4 +137,165 @@ private[events] object FilterTableACSReader {
...
+          val newIdsQueue = idsQueue.drop(1)
+          if (newIdsQueue.isEmpty) additionally.remove(task)
+          else additionally.addOne(task -> newIdsQueue)
+          (queueEntry(task, idsQueue.head), true)
+        case None =>
+          waitingFor.add(task)
+          (None, false)
+      }
+
+    private def queueEntry(task: TASK, ids: Iterable[Long]): Option[(Iterable[Long], TASK)] =
```
[the remark this answers is not in the document]
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala:294]: technically yes, but I would prefer to keep the option here to make this distinction explicit in the code, and especially between the components.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala [hunk 2]
```diff
@@ -135,4 +137,165 @@ private[events] object FilterTableACSReader {
...
+    *              - where each received batch should be exactly inputBatchSize-sized, except the last one
+    *              - the last one's size must not equal to inputBatchSize
+    * @param outputBatchSize The output stream will contain batches with this size, except the last one
+    * @param inputBatchSize batch size of the input
+    */
+  def mergeIdStreams[TASK](
+      tasks: Iterable[TASK],
+      outputBatchSize: Int,
+      inputBatchSize: Int,
+  ): () => ((TASK, Iterable[Long])) => IterableOnce[Vector[Long]] = () => {
```
[the remark this answers is not in the document]
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala:156]: Good point, quite some assumptions are here. I leave a note to the improvement task about this.

**Comment by nmarton-da** [comment]
> The logic is quite involved, if this is critical code then I think it needs more tests for edge cases and more precise comments. Otherwise it looks good, I have left a few cosmetic comments.

Well spotted: I've already created an improvement task to increase unit test coverage per component basis.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala [hunk 3]
```diff
@@ -135,4 +137,165 @@ private[events] object FilterTableACSReader {
...
+    /** Add one entry from upstream
+      * @param task the TASK identifier
+      * @param ids the ordered sequence of ids
+      * @return An optional entry to be added to the MergingTaskQueue, and a flag if further merging needed
+      */
+    def add(task: TASK, ids: Iterable[Long]): (Option[(Iterable[Long], TASK)], Boolean) = {
+      val toEnqueue =
+        if (waitingFor(task)) queueEntry(task, ids)
+        else {
+          additionally.addOne(task -> additionally.getOrElse(task, Vector.empty).:+(ids))
```
[the remark this answers is not in the document]
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala]: unfortunately not available in 2.12
